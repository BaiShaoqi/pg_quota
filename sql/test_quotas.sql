
--
CREATE USER quotatest_user NOLOGIN;

CREATE TABLE qt (t text);
ALTER TABLE qt OWNER TO quotatest_user;

INSERT INTO qt SELECT repeat('x', 100) FROM generate_series(1, 100000);

-- Wait a little, to make sure the background worker has picked up the new size.
select pg_sleep(5);

-- Display the table size.
select pg_size_pretty(pg_total_relation_size('qt'));

-- The "disk space used" as shown in quota_status should match
SELECT rolname,
       pg_size_pretty(space_used) as used,
       pg_size_pretty(quota) as quota
FROM quota.status
WHERE rolname::text like 'quotatest%';



-- Set a quota for the user.
INSERT INTO quota.config VALUES ('quotatest_user'::regrole, pg_size_bytes('20 MB'));

-- Wait a little, to give the worker a chance to pick up the new quota.
select pg_sleep(5);

SELECT rolname,
       pg_size_pretty(space_used) as used,
       pg_size_pretty(quota) as quota
FROM quota.status
WHERE rolname::text like 'quotatest%';

-- Now insert enough data that the quota is exceeded.
INSERT INTO qt SELECT repeat('x', 100) FROM generate_series(1, 100000);

-- and again wait a little, so that the worker picks up the new file size
select pg_sleep(5);

SELECT rolname,
       pg_size_pretty(space_used) as used,
       pg_size_pretty(quota) as quota
FROM quota.status
WHERE rolname::text like 'quotatest%';

-- Try to insert again. This should fail, because the quota is exceeded.
INSERT INTO qt SELECT repeat('x', 100) FROM generate_series(1, 100000);


-- Free up the space, by truncating the table. Now it should work again.
TRUNCATE qt;

-- and again wait a little, so that the worker picks up the new file size
select pg_sleep(5);

SELECT rolname,
       pg_size_pretty(space_used) as used,
       pg_size_pretty(quota) as quota
FROM quota.status
WHERE rolname::text like 'quotatest%';

INSERT INTO qt SELECT repeat('x', 100) FROM generate_series(1, 100000);
