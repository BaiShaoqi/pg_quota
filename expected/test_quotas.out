--
CREATE TABLE qt (t text);
INSERT INTO qt SELECT repeat('x', 100) FROM generate_series(1, 100000);
-- Display the table size.
select pg_size_pretty(pg_total_relation_size('qt'));
 pg_size_pretty 
----------------
 13 MB
(1 row)

-- Set a quota for the relation.
INSERT INTO quota.config VALUES ('qt'::regclass, pg_size_bytes('20 MB'));
-- Wait a little, to give the worker a chance to pick up the new quota.
select pg_sleep(5);
 pg_sleep 
----------
 
(1 row)

SELECT tablename,
       pg_size_pretty(space_used) as used,
       pg_size_pretty(quota) as quota
FROM quota.status;
 tablename | used  | quota 
-----------+-------+-------
 qt        | 13 MB | 20 MB
(1 row)

-- Now insert enough data that the quota is exceeded.
INSERT INTO qt SELECT repeat('x', 100) FROM generate_series(1, 100000);
-- and again wait a little, so that the worker picks up the new file size
select pg_sleep(5);
 pg_sleep 
----------
 
(1 row)

SELECT tablename,
       pg_size_pretty(space_used) as used,
       pg_size_pretty(quota) as quota
FROM quota.status;
 tablename | used  | quota 
-----------+-------+-------
 qt        | 26 MB | 20 MB
(1 row)

-- Try to insert again. This should fail, because the quota is exceeded.
INSERT INTO qt SELECT repeat('x', 100) FROM generate_series(1, 100000);
ERROR:  table's disk space quota exceeded
-- Free up the space, by truncating the table. Now it should work again.
TRUNCATE qt;
-- and again wait a little, so that the worker picks up the new file size
select pg_sleep(5);
 pg_sleep 
----------
 
(1 row)

SELECT tablename,
       pg_size_pretty(space_used) as used,
       pg_size_pretty(quota) as quota
FROM quota.status;
 tablename |    used    | quota 
-----------+------------+-------
 qt        | 8192 bytes | 20 MB
(1 row)

INSERT INTO qt SELECT repeat('x', 100) FROM generate_series(1, 100000);
